### DISCUSSION TITLE
Optional outsourcing of encryption and signature (with e.g. KMS)

### DISCUSSION THREAD

**Discussion by paulolieuthier** [keycloak/keycloak#22264] [Ideas]
We would like to use Keycloak to manage access internally. But we don't want to manage encryption keys. Ideally, encryption and signing would be outsourced to a service like AWS' KMS. The Key SPI as far as I can tell only allows different sources of keys, not using an external service.

Is that something you would be interested? With some guidance, I could work on that.

**Comment by vmuzikar** [comment]
@paulolieuthier Thanks for offering the contribution! Sounds like something we might be interested in. On the other hand, if I understand correctly (I'm no AWS expert) sounds like the cryptography is handled by an external system  (AWS KMS). Keycloak expects working with keys directly on many places. So it might not be something very straightforward.

CC @keycloak/core

**Reply by paulolieuthier** [reply]
Yes, cryptography would be handled externally. I took a deeper look at the source code and I saw how the keys are used directly in many different places. I think that ideally there would be an SPI for encryption operations.

**Reply by pedroigor** [reply]
Are you referring to PKCS#11 and the usage of crypto modules/devices/services?

I'm not sure about AWS KMS but if it supports PKCS#11 it is very likely we can integrate with it.

**Reply by paulolieuthier** [reply]
@pedroigor no, I'm referring to the regular token signing.

**Reply by stianst** [reply]
So, there's multiple layers, there's key providers, but also signing providers. To outsource signing to an external service or whatever you'd do a signing provider, but would also need to have the corresponding public key providers as those are needed for example in the jwks endpoints.

**Reply by paulolieuthier** [reply]
@stianst thanks for the clarification. But after digging into the source code it seems to me I would also have to provide the private key.

**Reply by stianst** [reply]
> @stianst thanks for the clarification. But after digging into the source code it seems to me I would also have to provide the private key.

Been quite a while since I've looked at that code, but that is at least the design around it. I'd say give it a go with a signing provider, without key provider first.

Worth mentioning that it may not work though as we don't have any signing providers that doesn't use key providers.

**Reply by paulolieuthier** [reply]
Looks like I reached a dead-end.

I implemented a `KeyProvider` which provides `KeyWrapper`s for every algorithm available from some configured KMS keys (KMS keys are configured through the key provider factory, for each realm) and `SignatureProvider`s for all Keycloak-supported algorithms (`HS256`, `HS384`, `HS512`, `RS256`, `RS384`, `RS512`, `ES256`, `ES384`, `ES512`, `PS256`, `PS384`, `PS512`), replacing the built-in providers.

Replacing the built-in signature providers is necessary, because they may selected by ID, it's hardcoded ([here](https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/jose/jws/DefaultTokenManager.java#L191-L195), [here](https://github.com/keycloak/keycloak/blob/main/server-spi-private/src/main/java/org/keycloak/models/Constants.java#L66) and [here](https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/jose/jws/DefaultTokenManager.java#L74-L79)).

Now I can't log in to the master realm, because there is no KMS key provider configured in the realm, and thus the KMS signature provider is failing. And I can't add the KMS key provider through system properties or environment variables, because it must be added manually to each realm. If I don't replace the built-in signature providers, I can log in to the master realm, but my HMAC signature provider is not selected to be used.

This feature is said to have been implemented [here](https://www.grootan.com/casestudies/to-provide-a-es-256-es-384-and-es-512-signing-way-for-access-token-of-open-id-with-aws-kms-instead-of-native-way-provided-by-keycloak), but AFAICT there is no source code available.

The possible solutions I can visualize:
* The signature providers could be configurable per realm, which would allow me to replace the built-in providers only on select realms.
* The key provider could be configurable globally or one could choose to add it to a realm via a system property or environment variable.

The last one would be preferred because I would like to configure this for all realms.

**Reply by paulolieuthier** [reply]
@stianst have you had the chance to take a look at this problem?
